Importing important library

In [61]:
import pandas as pd
import numpy as np
import warnings
import seaborn as sns
import matplotlib.pyplot as plt
warnings.filterwarnings("ignore")

pd.set_option("display.max_columns",None)

Loading the data

In [62]:
data = pd.read_csv(r'D:\Interproject\Creditcardfrauddetection\creditcard.csv')
data.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9       V10       V11       V12       V13       V14  \
0  0.098698  0.363787  0.090794 -0.551600 -0.617801 -0.991390 -0.311169   
1  0.085102 -0.255425 -0.166974  1.612727  1.065235  0.489095 -0.143772   
2  0.247676 -1.514654  0.207643  0.624501  0.066084  0.717293 -0.165946   
3  0.377436 -1.387024 -0.054952 -0.226487  0.178228  0.507757 -0.287924   
4 -0.270533  0.817739  0.753074 -0.822843  0.538196  1.345852 -1.119670   

        V15       V16       V17       V18       V19       V20       V21  \
0  1.468177 -0.470401  0.207971  0.025791  0.403993  0.251412 -0.018307   
1  0.635558  0.463917 -0.114805 -0.183361 -0.145783 -0.069083 -0.225775   
2  2.345865 -2.890083  1.109969 -0.121359 -2.261857  0.524980  0.247998   
3 -0.631418 -1.059647 -0.684093  1.965775 -1.232622 -0.208038 -0.108300   
4  0.175121 -0.451449 -0.237033 -0.038195  0.803487  0.408542 -0.009431   

        V22       V23       V24       V25       V26       V27       V28  \
0  0.277838 -0.110474  0.066928  0.128539 -0.189115  0.133558 -0.021053   
1 -0.638672  0.101288 -0.339846  0.167170  0.125895 -0.008983  0.014724   
2  0.771679  0.909412 -0.689281 -0.327642 -0.139097 -0.055353 -0.059752   
3  0.005274 -0.190321 -1.175575  0.647376 -0.221929  0.062723  0.061458   
4  0.798278 -0.137458  0.141267 -0.206010  0.502292  0.219422  0.215153   

   Amount  Class  
0  149.62      0  
1    2.69      0  
2  378.66      0  
3  123.50      0  
4   69.99      0

EDA

In [63]:
data.shape

(284807, 31)

There are 284,807 numbers of row and 31 columns

In [64]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

Maximum variable is in float datatype

In [65]:
data.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

There is no missing values present in the data

In [66]:
data.duplicated().sum()

1081

There are 1081 duplicate value which is 0.3 of whole data so we drop it

In [67]:
print('Before droping: ',data.shape)
data.drop_duplicates(inplace=True)
print('After droping: ',data.shape)

Before droping:  (284807, 31)
After droping:  (283726, 31)


In [68]:
colname = []
for x in data.columns:
    if data[x].dtype=='float':
        colname.append(x)

In [69]:
from sklearn.preprocessing import LabelEncoder

le = LabelEncoder()

for x in colname:
    data[x]=le.fit_transform(data[x])

Spliting the training and testing data

In [70]:
x = data.values[:,:-1]
y = data.values[:,-1]

In [71]:
print(x.shape)
print(y.shape)

(283726, 30)
(283726,)


In [72]:
from sklearn.preprocessing import StandardScaler

ss = StandardScaler()

ss.fit(x)
X = ss.transform(x)

In [73]:
y = y.astype(int)

In [74]:
from sklearn.model_selection import train_test_split

X_train, X_test, Y_train, Y_test = train_test_split(x, y, random_state=10,test_size=0.3)

In [75]:
print(X_train.shape)
print(X_test.shape)
print(Y_train.shape)
print(Y_test.shape)

(198608, 30)
(85118, 30)
(198608,)
(85118,)


Building Logistic Regression

In [76]:
from sklearn.linear_model import LogisticRegression

LR = LogisticRegression()

LR.fit(X_train,Y_train)

Y_pred = LR.predict(X_test)
print(Y_pred)

[0 0 0 ... 0 0 0]


In [77]:
Y_pred_prob = LR.predict_proba(X_test)
Y_pred_prob

array([[9.99982862e-01, 1.71379306e-05],
       [9.99891837e-01, 1.08162562e-04],
       [9.99995743e-01, 4.25694573e-06],
       ...,
       [9.99920445e-01, 7.95551697e-05],
       [9.99390422e-01, 6.09578198e-04],
       [9.99917052e-01, 8.29478888e-05]])

In [78]:
print(list(zip(Y_test,Y_pred)))

[(0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0),

In [79]:
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report
 
cfm=confusion_matrix(Y_test,Y_pred)
print(cfm)
 
print("Classification report: ")
 
print(classification_report(Y_test,Y_pred))
 
acc=accuracy_score(Y_test, Y_pred)
print("Accuracy of the model: ",acc)

[[84967    21]
 [   34    96]]
Classification report: 
              precision    recall  f1-score   support

           0       1.00      1.00      1.00     84988
           1       0.82      0.74      0.78       130

    accuracy                           1.00     85118
   macro avg       0.91      0.87      0.89     85118
weighted avg       1.00      1.00      1.00     85118

Accuracy of the model:  0.9993538382010856


The accuracy level is 99% but the type 2 error is greater than the correctly predicted value<br>
As we can see the class variable is imbalance, class 0 has 84988 out of 85118.<br>
Only 141 data for class 1 out of 85118<br>
We have to balance the data

Building the Decision Tree

In [80]:
from sklearn.tree import DecisionTreeClassifier
model_DecisionTree = DecisionTreeClassifier(random_state=10,criterion='gini')

model_DecisionTree.fit(X_train,Y_train)

DecisionTreeClassifier(random_state=10)

In [81]:
Y_pred = model_DecisionTree.predict(X_test)
print(Y_pred)
print(list(zip(Y_test,Y_pred)))

[0 0 0 ... 0 0 0]
[(0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0

In [82]:
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report
 
cfm=confusion_matrix(Y_test,Y_pred)
print(cfm)
 
print("Classification report: ")
 
print(classification_report(Y_test,Y_pred))
 
acc=accuracy_score(Y_test, Y_pred)
print("Accuracy of the model: ",acc)

[[84952    36]
 [   43    87]]
Classification report: 
              precision    recall  f1-score   support

           0       1.00      1.00      1.00     84988
           1       0.71      0.67      0.69       130

    accuracy                           1.00     85118
   macro avg       0.85      0.83      0.84     85118
weighted avg       1.00      1.00      1.00     85118

Accuracy of the model:  0.999071876688832


Building the Random Forest Classifier

In [83]:
from sklearn.ensemble import RandomForestClassifier
 
model_RandomForest=RandomForestClassifier(n_estimators=100,
                                          random_state=10, bootstrap=True,
                                         n_jobs=-1)
 
model_RandomForest.fit(X_train,Y_train)

RandomForestClassifier(n_jobs=-1, random_state=10)

In [84]:
Y_pred=model_RandomForest.predict(X_test)

In [85]:
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report
 
cfm=confusion_matrix(Y_test,Y_pred)
print(cfm)
 
print("Classification report: ")
 
print(classification_report(Y_test,Y_pred))
 
acc=accuracy_score(Y_test, Y_pred)
print("Accuracy of the model: ",acc)

[[84982     6]
 [   37    93]]
Classification report: 
              precision    recall  f1-score   support

           0       1.00      1.00      1.00     84988
           1       0.94      0.72      0.81       130

    accuracy                           1.00     85118
   macro avg       0.97      0.86      0.91     85118
weighted avg       1.00      1.00      1.00     85118

Accuracy of the model:  0.9994948189572124


As our data is not balanced so we have to balance the data<br>
We will perform the undersampling technique from resampling

In [86]:
data['Class'].value_counts()

Class
0    283253
1       473
Name: count, dtype: int64

In [87]:
from sklearn.utils import resample
# Separate majority and minority classes
df_majority = data[data.Class==0]
df_minority = data[data.Class==1]
 
# Upsample minority class
df_majority_upsampled = resample(df_majority, 
                                 replace=True,     # sample with replacement
                                 n_samples=500,    # to match majority class
                                 random_state=10) # reproducible results
 
# Combine majority class with upsampled minority class
df_upsampled = pd.concat([df_minority, df_majority_upsampled])
 
# Display new class counts
df_upsampled.Class.value_counts()

Class
0    500
1    473
Name: count, dtype: int64

In [88]:
X = df_upsampled.values[:,:-1]
Y = df_upsampled.values[:,-1]

In [89]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

scaler.fit(X)

X = scaler.transform(X)
#print(X)

In [90]:
from sklearn.model_selection import train_test_split

#Split the data into test and train
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.3,
                                                    random_state=10)  


In [91]:
print(X_train.shape)
print(X_test.shape)
print(Y_train.shape)
print(Y_test.shape)

(681, 30)
(292, 30)
(681,)
(292,)


In [92]:
# logistic regression object
lr = LogisticRegression()
  
# train the model on train set
lr.fit(X_train, Y_train)
  
Y_pred = lr.predict(X_test)

In [93]:
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report
 
cfm=confusion_matrix(Y_test,Y_pred)
print(cfm)
 
print("Classification report: ")
 
print(classification_report(Y_test,Y_pred))
 
acc=accuracy_score(Y_test, Y_pred)
print("Accuracy of the model: ",acc)

[[146   6]
 [ 11 129]]
Classification report: 
              precision    recall  f1-score   support

           0       0.93      0.96      0.94       152
           1       0.96      0.92      0.94       140

    accuracy                           0.94       292
   macro avg       0.94      0.94      0.94       292
weighted avg       0.94      0.94      0.94       292

Accuracy of the model:  0.9417808219178082


After undersampling of the data the Logistic Regression accuracy is 94%

In [94]:
from sklearn.tree import DecisionTreeClassifier
model_DecisionTree = DecisionTreeClassifier(random_state=10,criterion='gini')

model_DecisionTree.fit(X_train,Y_train)
Y_pred = model_DecisionTree.predict(X_test)

In [95]:
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report
 
cfm=confusion_matrix(Y_test,Y_pred)
print(cfm)
 
print("Classification report: ")
 
print(classification_report(Y_test,Y_pred))
 
acc=accuracy_score(Y_test, Y_pred)
print("Accuracy of the model: ",acc)

[[140  12]
 [ 16 124]]
Classification report: 
              precision    recall  f1-score   support

           0       0.90      0.92      0.91       152
           1       0.91      0.89      0.90       140

    accuracy                           0.90       292
   macro avg       0.90      0.90      0.90       292
weighted avg       0.90      0.90      0.90       292

Accuracy of the model:  0.9041095890410958


After undersampling of the data the Dicision Tree accuracy is 90%

In [96]:
from sklearn.ensemble import RandomForestClassifier
 
model_RandomForest=RandomForestClassifier(n_estimators=100,
                                          random_state=10, bootstrap=True,
                                         n_jobs=-1)
 
model_RandomForest.fit(X_train,Y_train)
Y_pred=model_RandomForest.predict(X_test)

In [97]:
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report
 
cfm=confusion_matrix(Y_test,Y_pred)
print(cfm)
 
print("Classification report: ")
 
print(classification_report(Y_test,Y_pred))
 
acc=accuracy_score(Y_test, Y_pred)
print("Accuracy of the model: ",acc)

[[148   4]
 [ 17 123]]
Classification report: 
              precision    recall  f1-score   support

           0       0.90      0.97      0.93       152
           1       0.97      0.88      0.92       140

    accuracy                           0.93       292
   macro avg       0.93      0.93      0.93       292
weighted avg       0.93      0.93      0.93       292

Accuracy of the model:  0.928082191780822


After undersampling of the data the Random Forest accuracy is 92%

After performing the undersampling the Logistic Regression is giving the best accuracy of 94% with that the type 2 error is also less compare to another model.